src/inference/Inference.jl

```diff
@@ -117,51 +117,38 @@ function assume(spl::A,
 
 end
 
-function assume(spl::A,
-    dists::Vector{T},
+function assume(spl::Union{SampleFromPrior, SampleFromUniform},
+    dists::AbstractVector{<:Distribution},
     vn::VarName,
-    var::Any,
-    vi::VarInfo) where {T<:Distribution, A<:Union{SampleFromPrior, SampleFromUniform}}
-
-    @assert length(dists) == 1 "Turing.assume only support vectorizing i.i.d distribution"
-    dist = dists[1]
-    n = size(var)[end]
+    vi::VarInfo)
 
+    n = length(dists)
     vns = map(i -> VarName(vn, "[$i]"), 1:n)
 
     if haskey(vi, vns[1])
         rs = vi[vns]
     else
-        rs = isa(spl, SampleFromUniform) ? init(dist, n) : rand(dist, n)
-
-        if isa(dist, UnivariateDistribution) || isa(dist, MatrixDistribution)
-            for i = 1:n
-                push!(vi, vns[i], rs[i], dist, spl)
-            end
-            @assert size(var) == size(rs) "Turing.assume: variable and random number dimension unmatched"
-            var = rs
-        elseif isa(dist, MultivariateDistribution)
-            for i = 1:n
-                push!(vi, vns[i], rs[:,i], dist, spl)
-            end
-            if isa(var, Vector)
-                @assert length(var) == size(rs)[2] "Turing.assume: variable and random number dimension unmatched"
-                for i = 1:n
-                    var[i] = rs[:,i]
-                end
-            elseif isa(var, Matrix)
-                @assert size(var) == size(rs) "Turing.assume: variable and random number dimension unmatched"
-                var = rs
-            else
-                @error("Turing.assume: unsupported variable container"); error()
-            end
+        if length(dists) == 1
+            dist = dists[1]
+            rs = isa(spl, SampleFromUniform) ? [init(dist) for i in 1:n] : [rand(dist) for i in 1:n]
+        elseif length(dists) == n
+            rs = isa(spl, SampleFromUniform) ? init.(dists) : rand.(dists)
+        else
+            throw("The distribution and values vectors have mis-matching lengths.")
         end
-    end
 
-    # acclogp!(vi, sum(logpdf_with_trans(dist, rs, istrans(vi, vns[1]))))
+        push!.(Ref(vi), vns, rs, dists, Ref(spl))
+        var = rs
+    end
```

Comment:
Do we still handle different formats of multiple samples from multivariate distribution with this change? E.g. `Vector{Vector}` and `Matrix`.

Reply:
Same question below.

Reply:
No the output of `vi[vns]` is now always a vector.

Reply:
Then how does it work with multivariate distribution?

Reply:
Like will `x ~ [Fill(MvNormal(), 2)]` works both for `x::Matrix` and `x::Vector{Vecotr}`?

Reply:
For `x::Matrix` that is column-wise, `~ MvNormal(zeros(2), 1)`, the user would need to do `x ~ MvNormal(zeros(2), 1)`. This is case 1 in the proposal given below https://github.com/TuringLang/Turing.jl/pull/819#issuecomment-504994714. This will be covered by the `dist::Distribution` method (to be changed in this PR) not the `dist::Vector{<:Distribution}` method which this PR currently modifies.

`x ~ [Fill(MvNormal(zeros(2), 1), 2)]` should arguably error. The rhs is a `Vector{<:Vector{<:Distribution}}`. I think we can should only support `Array{<:Distribution}`. This can be through `map` semantics  using `lhs ~ rhs::Array{<:Distribution}` where `size(lhs) == size(rhs)` (otherwise error), case 3 in the proposal below. Or it can be through broadcast semantics using `.~` which broadcasts `lhs` and `rhs` to matching sizes, case 2 in the proposal below.